File: exercism/python/wordy/wordy.py
"""
wordy.py
"""

import re


def answer(question):
    """
    Parse and evaluate simple math word problems returning the answer as an integer.
    """

    # validate syntax
    if not question.startswith('What is'):
        raise ValueError("unknown operation")

    op_regex = re.compile(r'(plus|minus|divided|multiplied)')
    num_regex = re.compile(r'-?\d+')

    operations = list(re.findall(op_regex, question))
    numbers = list(re.findall(num_regex, question))

    # returns error if numbers and operators mismatch
    if len(operations) != (len(numbers) - 1):
        raise ValueError("syntax error")

    if (not re.match(r"What is \d\?", question)) and len(numbers) == 1:
        raise ValueError("unknown operation")

    if not re.match(r"What is -?\d+(?: (?:plus|minus|divided by|multiplied by) -?\d+)*\?",
                    question):
        raise ValueError("syntax error")

    total = int(numbers[0])

    for index, operation in enumerate(operations):
        number = int(numbers[index+1])
        if operation == "plus":
            total += number
        elif operation == "minus":
            total -= number
        elif operation == "divided":
            total /= number
        elif operation == "multiplied":
            total *= number

    return total

File: exercism/go/wordy/wordy.go
package wordy

import (
	"regexp"
	"strconv"
)

func Answer(question string) (int, bool) {

	// validate syntax
	if match, _ := regexp.MatchString(`What is -?\d+(?: (?:plus|minus|divided by|multiplied by) -?\d+)*\?`, question); !match {
		return 0, false
	}

	// gets the opetators and integers in a slice
	re1 := regexp.MustCompile(`(plus|minus|divided|multiplied)`)
	operators := re1.FindAllString(question, -1)
	re2 := regexp.MustCompile(`-?\d+`)
	numbers := re2.FindAllString(question, -1)

	// returns error if numbers and operators mismatch
	if len(operators) != len(numbers)-1 {
		return 0, false
	}
	sum, _ := strconv.Atoi(numbers[0])
	for i, o := range operators {
		n, _ := strconv.Atoi(numbers[i+1])
		switch o {
		case "plus":
			sum += n
		case "minus":
			sum -= n
		case "divided":
			sum /= n
		case "multiplied":
			sum *= n
		}
	}
	return sum, true
}
